### ISSUE TITLE
Excluding files unexpected behavior.

### ISSUE THREAD

**Issue by AIGeneratedUsername** [crate-ci/typos#347]
**Current settings:**

__typos.toml_
```toml
[files]
extend-exclude = ["*.json", "**/*.json"]
```

_.pre-commit-config.yaml_
```
-   repo: https://github.com/crate-ci/typos
    rev: typos-v0.7.4
    hooks:
      - id: typos
```

**Expected behavior:**
JSON files must be excluded 

**Expected behavior:**
All files are checked.

**Steps to reproduce**

` pre-commit run typos --all-files`

**Comment by epage** [comment]
`typos` assumes the user knows best when passing in file paths, overriding the exclusions.  Unfortunately, this doesn't layer well with tools like pre-commit.

We need to provide a way to control this behavior.

**Comment by amitlevy21** [comment]
I have the same issue, but only when running `typos-src` with `pre-commit`, running typos from the terminal works fine.
this is my config (_typos.toml at project root directory):
```toml
[files]
extend-exclude = ["src/assets/locales"]

```

locale files are ignored when running from the terminal, but not from `pre-commit run --all-files -v typos-src`

```yaml
  - repo: https://github.com/crate-ci/typos
    rev: v1.3.3
    hooks:
      - id: typos-src
        # args:
        #   - --exclude
        #   - src/assets/locales #uncommenting args works, but its not reading my config file

```

**Comment by kwanhur** [comment]
> `typos` assumes the user knows best when passing in file paths, overriding the exclusions. Unfortunately, this doesn't layer well with tools like pre-commit.
> 
> We need to provide a way to control this behavior.

What about the hook attribute [exclude](https://pre-commit.com/#hooks-exclude) ? It can exclude files with specified pattern.

**Comment by TomaszAIR** [comment]
Hi, we have a little bit different situation

typos.toml

```
$ cat typos.toml 
[files]
extend-exclude = ["Cargo.lock"]
```
.pre-commit-config.yaml

```
repos:
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v4.1.0
    hooks:
      - id: check-added-large-files
      - id: check-merge-conflict
      - id: check-symlinks
      - id: detect-private-key
      - id: end-of-file-fixer
      - id: trailing-whitespace
  - repo: https://github.com/crate-ci/typos
    rev: v1.4.1
    hooks:
      - id: typos
```

Running couple of times `pre-commit run typos --all-files` gives different results. 2 passes and 1 fail

```
$ pre-commit run --all-files                  
check for added large files..............................................Passed
check for merge conflicts................................................Passed
check for broken symlinks............................(no files to check)Skipped
detect private key.......................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
typos....................................................................Passed
$ pre-commit run --all-files                  
check for added large files..............................................Passed
check for merge conflicts................................................Passed
check for broken symlinks............................(no files to check)Skipped
detect private key.......................................................Passed     
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
typos....................................................................Passed
$ pre-commit run --all-files
check for added large files..............................................Passed
check for merge conflicts................................................Passed                        
check for broken symlinks............................(no files to check)Skipped           
detect private key.......................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
typos....................................................................Failed
```

After that Cargo.lock was changed

```
$ git diff Cargo.lock
diff --git a/Cargo.lock b/Cargo.lock
index 938bfc3b47f7..68cca07e67e5 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -296,7 +296,7 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "d1873270f8f7942c191139cb8a40fd228da6c3fd2fc376d7e92d47aa14aeb59e"
 dependencies = [
  "crypto-common",
- "inout",
+ "input",
 ]
 
 [[package]]
@@ -769,7 +769,7 @@ dependencies = [
 ]
 
 [[package]]
-name = "inout"
+name = "input"
 version = "0.1.2"
```

As if exclusions from `typos.toml` weren't applied every time. Has anyone encountered a similar problem? It looks like a similar case that was here https://github.com/crate-ci/typos/issues/278

**Comment by epage** [comment]
> typos.toml
> 
> ```
> $ cat typos.toml 
> [files]
> extend-exclude = ["Cargo.lock"]
> ```

This is making me realize we should probably centralize all locks into a single file type and have it excluded.

**Comment by epage** [comment]
@TomaszAIR v1.5.0 was just released where we do not check lock files.  This is done in a different way than `extend-exclude` and should work with pre-commit.

**Comment by TomaszAIR** [comment]
@epage thanks for quick fix. Will test that and let you know how it goes.

**Comment by TomaszAIR** [comment]
Yeap, it works. Can confirm that v1.5.0 do not check lock files. None of the `extend-exclude` configuration is needed. Thanks

**Comment by sshishov** [comment]
Hello, when we can expect the fix for the issue? Just a question, not pushing for any actions 😉

**Comment by epage** [comment]
> Hello, when we can expect the fix for the issue?

The core of this is we need a natural mode where we trust the user and a scripted mode where we don't trust the user.  The blocker is defining how to express this in the UI.   Bonus points for including any precedence from other tools

Also want to note that #461 has the workaround of duplicating the exclusions

**Comment by sshishov** [comment]
Thanks @epage , currently as a workaround we started using `pre-commit` exclude, but added `#TODO:` with appropriate comment. And also keep the configuration to exclude in the config file

**Comment by sshishov** [comment]
Additional workaround is:
```
  - repo: https://github.com/crate-ci/typos
    rev: v1.8.1
    hooks:
    - id: typos
      pass_filenames: false
```
Got it from [here](https://github.com/econchick/interrogate/issues/60#issuecomment-735436566) as the same issue exists in [interrogate](https://github.com/econchick/interrogate) package

**Comment by sshishov** [comment]
The same issue as in [isort](https://github.com/PyCQA/isort/issues/938) fixed by [PR 939](https://github.com/PyCQA/isort/pull/939) and [black](https://github.com/psf/black/issues/438) fixed by [PR 1032](https://github.com/psf/black/pull/1032)

Same problem was happening with `flake8` and happening now with [interrogate][(https://github.com/crate-ci/typos/issues/347)](https://github.com/econchick/interrogate/issues/60)

**Comment by epage** [comment]
@sshishov thanks for those examples.  To summarize:
- `isort` added `--filter-files`
- `black` added `--force-exclude`

I wish there was a flag that was common between tools, even if the sample size is just two, as being consistent can be more important than being ideal on its own.

Since these aren't consistent, my next thought is on which more clearly communicates intent.  I feel like isort's only makes sense when reading the docs and thinking about what it is trying to say.  black's makes a little more sense.  "force" make it clear this is a binary flag and that it is overriding some default behavior.  I lean towards that somewhat even though something still feels off to me. Really, we are implicitly forcing the the args to be respected independent of the exclusions.  I just can't think of a succinct way of saying that, so maybe just mirroring black would be best.

If people know of other examples, I'm still open open on this.  At least for me adding the flag, it might be a week or so as I'm trying to wrap up some major work on the clap crate.

**Comment by adamchainz** [comment]
> What about the hook attribute [exclude](https://pre-commit.com/#hooks-exclude) ? It can exclude files with specified pattern.

This is what I'm using. For example:

```yaml
repos:
- repo: https://github.com/crate-ci/typos
  rev: v1.13.10
  hooks:
    - id: typos
      exclude: |
        (?x)^(
          .*\.csv
          |.*\.xml
        )$
```

You can also explicitly limit to certain file types by overriding the `text` type selected in the repo's hook definition:

```yaml
repos:
- repo: https://github.com/crate-ci/typos
  rev: v1.13.10
  hooks:
    - id: typos
      types_or:
      - javascript
      - python
```

IMO one of the powers of pre-commit is that it already passes the correct filenames, based on hook selection and exclusion, so tools don't have to reimplement undifferentiated inclusion/exclusion features.

**Comment by epage** [comment]
> IMO one of the powers of pre-commit is that it already passes the correct filenames, based on hook selection and exclusion, so tools don't have to reimplement undifferentiated inclusion/exclusion features.

For tools written exclusively for pre-commit, that works well.  However, this tool is written more generally and needs to have its own logic.

**Comment by Delgan** [comment]
Hi @epage.

Would you accept a PR adding a new `--force-exclude` similar to the one used by `black`?

> **`--force-exclude`**
> Like `--exclude`, but files and directories matching this regex will be excluded even when they are passed explicitly as arguments. This is useful when invoking Black programmatically on changed files, such as in a pre-commit hook or editor plugin.

I really wish I could use `typos` with `pre-commit` without the `--all-files` workaround.

**Comment by epage** [comment]
Yes, when looking at black before, I thought it was a bool and not yet another exclude type which seems like it'd be awkward to deal with.

> I think by default, we want typos to ignore excluded files even if they're explicitly passed to the command line. This is because using typos *.py is a very common usage, so it makes sense that exclude takes precedence.

Is `typos *.py` very common?  For myself, I run it on directories or on individual files and neither case runs into this problem.  

If we went this route, it'd be a breaking change and we'd need to bump to 2.0.  Not the end of the world but something to keep in mind.

I also feel like we'd need a way to help the user discover that we ignored what they said and they need to pass in an extra flag.  Not thrilled with doing that.

Most of this problem seems to be centered on pre-commit-like workflows where we shouldn't be trusting the external source of information.

**Comment by Delgan** [comment]
I've done more research and compared other tools, and I admit that what I've suggested isn't that common. It depends on people's preferences, but there doesn't seem to be any consensus.
Personally, I'm more likely to use `typos *.py` than to call `typos` on a file I've explicitly configured to be excluded.

Just for example, there is **ESLint** which displays a warning when excluded files are passed to the command line and requires an explicit `--no-ignore`: [Ignore Files - ESLint - Pluggable JavaScript Linter](https://eslint.org/docs/latest/use/configure/ignore).

However if we take into account the issue regarding backwards compatibility, I agree it's definitely preferable not to change the existing behavior. Apologies for not thinking of that.

As a further example, there is also **Ruff** which implemented `--force-exclude` as a boolean flag and directly integrated it to their pre-commit hook: [`ruff-pre-commit/.pre-commit-hooks.yaml#L4`](https://github.com/astral-sh/ruff-pre-commit/blob/ba4269e9c6827628d9ff4732249ba22fbde46ae7/.pre-commit-hooks.yaml#L4).
This matches with solutions discussed in this thread and we could copy it for a touch of consistency.

I can understand your hesitations to add such a flag, but its primary purpose is to enable `typos` to be used programmatically. Not only for pre-commit, but also to any Bash script or VS Code extension that wants to use `typos` without the burden of maintaining its own exclude list.
From my perspective, it's more of a valuable feature than a mere workaround.

I am always willing to open a PR if you wish.

**Comment by epage** [comment]
My concern with `--force-exclude` was more centered on black having it be a separate list of excludes.  Having it be a bool would make sense and I'd be willing to be consistent with ruff in this.

**Comment by epage** [comment]
Closed via #837
